presto-orc/src/main/java/com/facebook/presto/orc/writer/SliceDictionaryColumnWriter.java

```diff
@@ -194,15 +222,22 @@ public long convertToDirect()
         return directColumnWriter.getBufferedBytes();
     }
 
-    private void writeDictionaryRowGroup(Block dictionary, int valueCount, IntBigArray dictionaryIndexes)
+    private void writeDictionaryRowGroup(Block dictionary, int valueCount, LongInputStream dictionaryIdInputStream)
     {
-        int[][] segments = dictionaryIndexes.getSegments();
-        for (int i = 0; valueCount > 0 && i < segments.length; i++) {
-            int[] segment = segments[i];
-            int positionCount = Math.min(valueCount, segment.length);
-            DictionaryBlock dictionaryBlock = new DictionaryBlock(positionCount, dictionary, segment);
+        int[] dictionaryIdBuffer = new int[1024];
+
+        while (valueCount > 0) {
+            int batchSize = Math.min(valueCount, dictionaryIdBuffer.length);
+            try {
+                dictionaryIdInputStream.nextIntVector(batchSize, dictionaryIdBuffer, 0);
+            }
+            catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+
+            DictionaryBlock dictionaryBlock = new DictionaryBlock(batchSize, dictionary, dictionaryIdBuffer);
```

Comment:
The `id`s array can not be reused for multiple dictionary blocks.  The block does not copy the data.  In this case it likely works because the callee does not retain the block, but I'd rather be safe.